src/librustc_passes/check_const.rs

```diff
@@ -135,15 +152,47 @@ impl<'tcx> CheckConstVisitor<'tcx> {
 
         let const_kind = self.const_kind
             .expect("`const_check_violated` may only be called inside a const context");
-
         let msg = format!("`{}` is not allowed in a `{}`", expr.name(), const_kind);
-        match expr {
-            | NonConstExpr::Match(hir::MatchSource::Normal)
-            | NonConstExpr::Match(hir::MatchSource::IfDesugar { .. })
-            | NonConstExpr::Match(hir::MatchSource::IfLetDesugar { .. })
-            => feature_err(&self.tcx.sess.parse_sess, sym::const_if_match, span, &msg).emit(),
 
-            _ => span_err!(self.tcx.sess, span, E0744, "{}", msg),
+        let gates = gates.unwrap_or(&[]);
+        let missing_gates: Vec<_> = gates
+            .iter()
+            .copied()
+            .filter(|&g| !features[g])
+            .collect();
+
+        match missing_gates.as_slice() {
+            &[] => span_err!(self.tcx.sess, span, E0744, "{}", msg),
+
+            // If the user enabled `#![feature(const_loop)]` but not `#![feature(const_if_match)]`,
+            // explain why their `while` loop is being rejected.
+            &[gate @ sym::const_if_match] if gates.contains(&sym::const_loop) => {
+                let mut err = feature_err(&self.tcx.sess.parse_sess, gate, span, &msg);
+                err.note("`#![feature(const_loop)]` alone is not sufficient, \
+                          since this loop expression contains an implicit conditional");
+                err.emit();
+            }
+
+            &[missing_primary, ref missing_secondary @ ..] => {
+                let mut err = feature_err(&self.tcx.sess.parse_sess, missing_primary, span, &msg);
+
+                // If multiple feature gates would be required to enable this expression, include
+                // them as help messages. Don't emit a separate error for each missing feature gate.
+                //
+                // FIXME(ecstaticmorse): Maybe this could be incorporated into `feature_err`? This
```

Comment:
I would prefer just emitting two gate errors instead to simplify the logic and ensure its correctness.

Reply:
I disagree. I think the first is better for UX.

Current approach:

```text
error[E0658]: `while` is not allowed in a `const`
  --> $DIR/loop.rs:51:5
   |
LL | /     while x < 8 {
LL | |         x += 1;
LL | |     }
   | |_____^
   |
   = note: for more information, see https://github.com/rust-lang/rust/issues/52000
   = help: add `#![feature(const_loop)]` to the crate attributes to enable
   = help: add `#![feature(const_if_match)]` to the crate attributes to enable
```

Multiple errors

```text
error[E0658]: `while` is not allowed in a `const`
  --> $DIR/loop.rs:51:5
   |
LL | /     while x < 8 {
LL | |         x += 1;
LL | |     }
   | |_____^
   |
   = note: for more information, see https://github.com/rust-lang/rust/issues/52000
   = help: add `#![feature(const_loop)]` to the crate attributes to enable

error[E0658]: `while` is not allowed in a `const`
  --> $DIR/loop.rs:51:5
   |
LL | /     while x < 8 {
LL | |         x += 1;
LL | |     }
   | |_____^
   |
   = note: for more information, see https://github.com/rust-lang/rust/issues/49146
   = help: add `#![feature(const_if_match)]` to the crate attributes to enable
```

Reply:
The complexity is not that large and this is indeed nice UX. Also it will go away with stabilization, so...

Reply:
Sure, it's undoubtedly better UX to emit one error. However, better UX should not be the primary concern with feature gating but rather that it is done in a correct manner. Having more special cases in the feature gating code makes this harder. Indeed, this code here does not account for `!span.allows_unstable($name)` as done in the `gate_feature_fn`. But it seems you feel strongly about keeping this so let's.